### PR TITLE
[TRTLLM-10061][feat] Prefix caching support for mamba hybrid models

### PR DESCRIPTION
**Pull request by VALLIS-NERIA**
## Release Notes

* **Bug Fixes**
  * Fixed stale cache block relationships that could occur in hybrid cache management
  * Improved Mamba state initialization to correctly handle initial state conditions

* **New Features**
  * Added block reuse capability for hybrid linear attention models with configurable state cache intervals
  * Enhanced KV cache management strategies for improved memory efficiency

* **Performance Improvements**
  * Optimized cache allocation and state management for Mamba-based models
  * Better memory utilization through improved cache block handling

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by VALLIS-NERIA** [comment]
/bot run --stage-list "A10-PyTorch-2, DGX_B200-PyTorch-1, DGX_B200-PyTorch-2, DGX_H100-PyTorch-1, DGX_H100-PyTorch-3, L40S-PyTorch-1"

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 1]
```diff
@@ -305,8 +305,8 @@ def _prepare_mamba_metadata(self):
             return
 
         if self.mamba_metadata is None:
-            if (self.kv_cache_manager is not None
-                    and isinstance(self.kv_cache_manager, MambaCacheManager)):
+            if (self.kv_cache_manager is not None and isinstance(
+                    self.kv_cache_manager, BaseMambaCacheManager)):
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/attention_backend/interface.py]: Nit: I think we can simplify to just `isinstance(self.kv_cache_manager, BaseMambaCacheManager)` since if it passes, it means `self.kv_cache_managers is not None`.

**Review by xinhe-nv — APPROVED** [review]

**Review by pengbowang-nv — APPROVED** [review]
Attention Part LGTM.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 2]
```diff
@@ -5935,66 +5943,6 @@ def test_nvfp4(self, moe_backend, tp_size, pp_size, ep_size, cuda_graph,
             task = GSM8K(self.MODEL_NAME)
             task.evaluate(llm)
 
-    @skip_pre_blackwell
-    @pytest.mark.skip_less_device(2)
-    def test_bf16_2gpu_mtp_ar(self):
```
Comment (eopXD) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py]: What is the replacement for deleting this test?

**Review by eopXD — COMMENTED** [review]
Minor questions and comments only. The enablement itself looks good to me.

The previous MR made this MR very clean.

File: cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp [hunk 3]
```diff
@@ -2093,6 +2093,16 @@ std::pair<SizeType32, std::vector<KVCacheBlock::IdType>> WindowBlockManager::sto
                 }
                 block->setBlockKey(blockKey, static_cast<SizeType32>(blockKey.uniqueTokens.size()) == mTokensPerBlock);
                 block->setPrevBlockInSeq(searchRoot);
+                if (!needMatch)
+                {
+                    auto const& rootNexts = searchRoot->getNextBlocks();
+                    if (rootNexts.find(blockKey) != rootNexts.end() && rootNexts.at(blockKey) != block)
+                    {
+                        // If blockKey has been a child, addNextBlock will have no effect. This may happen on
```
[the remark this answers is not in the document]
Comment (VALLIS-NERIA) [reply cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp]: It's some kind of noncritical bug. The expected behavior is that `searchRoot` either has no child at all, or has `block` as a child. But with mamba block reuse enabled, sometimes `searchRoot` has a child that is not `block`. According to the design of `addNextBlock`, it's a no-op if a child already presents. So `block` is not inserted to the tree and mess up the remaining process.

This snippet is a workaround by resetting it back to the expected status.

**Review by eopXD — APPROVED** [review]
The merge request looks good to me.

**Review by nv-guomingz — APPROVED** [review]

File: tensorrt_llm/_torch/modules/mamba/gdn_mixer.py [hunk 4]
```diff
@@ -763,9 +763,13 @@ def forward(
 
         state_indices_p, state_indices_d = torch.split(state_indices, batch_split_size)
         if num_prefills > 0:
-            ssm_states[state_indices_p] = torch.zeros(
+            has_initial_states_p = has_initial_states[:num_prefills]
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/modules/mamba/gdn_mixer.py]: Potentially stupid question: how do we know the first `num_prefills` elements are for prefill? Could we leave a comment for why that assumption is valid?

**Review by tcherckez-nvidia — APPROVED** [review]
LGTM, reviewed only AD changes

**Review by 2ez4bz — APPROVED** [review]

File: tensorrt_llm/_torch/model_config.py [hunk 5]
```diff
@@ -796,10 +806,32 @@ def get_layer_types(self) -> Optional[List[LayerTypeCpp]]:
         else:
             return None
 
-    def get_num_attention_layers(self):
-        if is_nemotron_hybrid(self.pretrained_config):
+    def get_num_attention_layers(
+            self,
+            kv_cache_config: Optional[KvCacheConfig] = None,
+            spec_config: Optional['SpeculativeConfig'] = None):
+        """Return the number of layers that need KV cache blocks.
```
Comment (Funatiq) [inline tensorrt_llm/_torch/model_config.py:813]: Isn't the function name misleading then? It should probably renamed to something like `get_num_cache_layers`.

**Review by Funatiq — APPROVED** [review]

File: tensorrt_llm/_torch/model_config.py (hunk 5 above)
Comment (VALLIS-NERIA) [reply tensorrt_llm/_torch/model_config.py:813]: This makes sense.  But changing one single line would take another ~5 attempts of CI to get over all those intermittent failures so I'd fix this in the coming PRs.

**Review by HuiGao-NV — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 6]
```diff
@@ -732,3 +803,432 @@ def update_mamba_states(self, attn_metadata: "AttentionMetadata",
...
+
+
+class CppMambaHybridCacheManager(KVCacheManager, BaseMambaCacheManager):
+    """Hybrid cache manager storing mamba states inside the KVCacheManager pool.
+
+    Both KV cache blocks and recurrent state blocks are managed by the unified
+    C++ KVCacheManager, enabling block reuse / prefix caching across attention
+    and mamba layers. This is the default hybrid manager.
+
+    Disaggregated serving and speculative decoding are not supported yet.
```
Comment (pcastonguay) [inline tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:835]: @VALLIS-NERIA what's missing to support disaggregated serving?

Comment (VALLIS-NERIA) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:835]: The mamba state uses layer-first layout, meaning the blocks are stride in memory. The CacheTransceiver does not yet support this kind of layout.

Comment (pcastonguay) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:835]: @bo-nv @NVShreyas can you confirm? I thought we supported Mamba models with disagg. What is the gap?

Comment (VALLIS-NERIA) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:832]: Current support of disagg is based on RnnStateManager and doesn't support kv cache reuse.

Comment (bo-nv) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:832]: @VALLIS-NERIA  these codes only work for CppMambaCachaManager? Currently, PythonMambaCacheManager also supports disagg, and we plan to deprecate CppMambaCachaManger in the feature.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 7]
```diff
@@ -58,8 +64,56 @@ def use_cpp_mamba_cache_manager() -> bool:
...
+               'PythonMambaCacheManager.SpeculativeState', None]:
+        ...
+
+
 class CppMambaCacheManager(BaseResourceManager):
-    """C++ backed Mamba cache manager using RnnStateManager bindings."""
+    """Mamba state manager backed by the C++ RnnStateManager bindings.
+
+    Manages only mamba states (conv + SSM). Used when TRTLLM_USE_CPP_MAMBA=1,
+    which is required for disaggregated serving deployments.
```
Comment (bo-nv) [inline tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: This is not reuired for disagg, as PythonMambaCacheManager also supports disagg.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 8]
```diff
@@ -499,7 +558,13 @@ def update_mamba_states(self, attn_metadata: "AttentionMetadata",
         conv_states[:, state_indices_d, :] = accepted_conv_state
 
 
-class MambaCacheManager(BaseResourceManager):
+class MambaCacheManager(BaseResourceManager, BaseMambaCacheManager):
+    """Facade for standalone mamba state management (no KV cache).
+
+    Delegates to CppMambaCacheManager (when TRTLLM_USE_CPP_MAMBA=1, required
+    for disaggregated serving) or PythonMambaCacheManager (default, supports
```
Comment (bo-nv) [inline tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: PythonMambaCacheManager also supports disagg

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 9]
```diff
@@ -732,3 +803,432 @@ def update_mamba_states(self, attn_metadata: "AttentionMetadata",
...
+              mamba_state_cache_interval))
+    last_ckpt = prompt_len // tokens_per_block * tokens_per_block
+    if save_last_snapshot and (last_ckpt not in stop_positions):
+        stop_positions.append(last_ckpt)
+    if prompt_len not in stop_positions:
+        stop_positions.append(prompt_len)
+    return stop_positions
+
+
+class CppMambaHybridCacheManager(KVCacheManager, BaseMambaCacheManager):
```
Comment (bo-nv) [inline tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: Is it possible to use PythonMambaCacheManager here, I think it supports both disagg and speculative decoding

`class CppMambaHybridCacheManager(KVCacheManager, PythonMambaCacheManager):`

Comment (bo-nv) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: With PythonMambaCacheManager, I think we just need one MambaHybridCacheManager?

Comment (VALLIS-NERIA) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: MixedMambaHybridCacheManager = two managers, one for mamba state (implemented in Python or CPP) and one for KV Cache
CppMambaHybridCacheManager = one single manager to manage both mamba state and kv cache

Comment (VALLIS-NERIA) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: So we will have three choices:
1. Single Cpp manager to manage all (support block reuse)
2. Python mamba manager + individual kv cache manager (support MTP and disagg)
3. Cpp mamba manager + individual kv cache manager (support disagg)

Do you mean the option 3 is useless now?

Comment (bo-nv) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: Yes, the option 3 will be deprecated in the future, I just don't understand why we need another c++ manager to support block reuse rather than extending the option.2, the new one is also derive from KVCacheManager + BaseMambaCacheManager

Comment (bo-nv) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: CC @Shixiaowei02

Comment (bo-nv) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: Oh, I see, the option.1 manages blocks in different way to support block reuse, so currently for mamba models, trtllm either supports block reuse or support disagg/mtp.

Comment (VALLIS-NERIA) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:825]: Yes. `BaseMambaCacheManager` is only an interface.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py (hunk 6 above)
Comment (VALLIS-NERIA) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py:832]: I think we are converged with function gap and backed implementations of these three variants. I'm going to only update the docstrings.

**Comment by farazkh80** [comment]
Hey @VALLIS-NERIA thanks for adding prefix support for Mamba. QQ, have you checked functionality on concurrency > 1 ?

I tried Nemo Omni models with 32k prefix cache, and conc1 works as expected TTFT drops by 5x, however conc 2 not so much => prefix_cache_c2 = 21991 ms vs prefix_cache_c1 = 904 ms.
 
```
[kv cache manager] storeContextBlocks: Can not find sequence for request 9
[kv cache manager] storeContextBlocks: Can not find sequence for request 8
 ```

**Comment by VALLIS-NERIA** [comment]
> Hey @VALLIS-NERIA thanks for adding prefix support for Mamba. QQ, have you checked functionality on concurrency > 1 ?
> 
> I tried Nemo Omni models with 32k prefix cache, and conc1 works as expected TTFT drops by 5x, however conc 2 not so much => prefix_cache_c2 = 21991 ms vs prefix_cache_c1 = 904 ms.
> 
> ```
> [kv cache manager] storeContextBlocks: Can not find sequence for request 9
> [kv cache manager] storeContextBlocks: Can not find sequence for request 8
> ```

@farazkh80 I have integration tests for Nemotron V3 on B200. The concurrency is ~48 and accuracy is fine (MMLU and GSM8K). Since currently it's based on chunked prefill, you can expect it's host bottleneck for prefill stage.
